### PR TITLE
Add sudoku

### PR DESCRIPTION
**Pull request by raskovnik**
Add sudoku on backtracking algorithm

### CODE REVIEW

File: src/backtracking/sudoku.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    }
+    return false;
+}
+
+pub fn is_in_box(board: &[[i8;9];9], row: &usize, col: &usize, number: &i8) -> bool {
+    let l_col = *&col - (*&col % 3);
+    let l_row = *&row - (*&row % 3);
+    for row in l_row..l_row + 3 {
+        for col in l_col..l_col + 3 {
+            if board[row][col] == *number {return true;}
```
Comment (st3v3nmw) [inline src/backtracking/sudoku.rs]: For readability, we could have the if as a full block, like return and closing bracket on separate lines

File: src/backtracking/sudoku.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+    return false;
+}
+
+pub fn is_valid_placement(board: &[[i8;9];9], row: &usize, col: &usize, number: &i8) -> bool {
+    if !is_in_box(board, row, col, number) && !is_in_column(board, col, number) && !is_in_row(board, row, number) {return true;}
+    else {return false;}
```
Comment (st3v3nmw) [inline src/backtracking/sudoku.rs]: Here as well

File: src/backtracking/sudoku.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        for j in 0..9 {
+            if board[i][j] == 0 {
+                for num in 1..=9 {
+                    if is_valid_placement(&board, &i, &j, &num) {
+                        board[*&i][*&j] = num;
+                        if solve(&mut *board) == false {
+                           board[*&i][*&j] = 0;
+                           empties += 1;
+                       }
+                       else {return true;}
```
Comment (st3v3nmw) [inline src/backtracking/sudoku.rs]: And here

File: src/backtracking/sudoku.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                        [0,0,0,0,0,3,0,0,0],
+                        [1,0,0,0,0,9,5,0,0],
+                        [8,0,0,0,0,0,0,9,0],
+                        [0,4,3,0,0,0,7,5,0],
+                        [0,9,0,0,0,0,0,0,8],
+                        [0,0,9,7,0,0,0,0,5],
+                        [0,0,0,2,0,0,0,0,0],
+                        [0,0,7,0,4,0,2,0,3]];
+
+        assert_eq!(solve(&mut board), false);
```
Comment (st3v3nmw) [inline src/backtracking/sudoku.rs]: You could add 1 more test for a solvable board

**Review by st3v3nmw — CHANGES_REQUESTED** [review]
Looking good so far! Work on the small issues raised

File: src/backtracking/sudoku.rs (hunk 4 above)
Comment (st3v3nmw) [reply src/backtracking/sudoku.rs]: Also, are we able to return the boolean & Option(Board) 😅? `(false, None)` for unsolved boards, & `(true, Some(Board))` ndo we be able to print later _outside_ the function.

It's still fine of you want that `solve` function to just return the bool

Comment (raskovnik) [reply src/backtracking/sudoku.rs]: it returns the bool but we're modifying the array. We'd just have to iterate over the array to print out the solution

Comment (st3v3nmw) [reply src/backtracking/sudoku.rs]: > it returns the bool but we're modifying the array. We'd just have to iterate over the array to print out the solution

Ah, I see, nice

**Pull request merged**